src/async_fuse/memfs/s3_metadata.rs

```diff
@@ -665,94 +656,137 @@ impl<S: S3BackEnd + Sync + Send + 'static> MetaData for S3MetaData<S> {
+            );
+            txn.set(
+                &KeyType::INum2Node(new_ino),
+                &ValueType::Node(new_node.into_serial_node()),
+            );
+
+            if let Some(mut new_parent_node) = new_parent_node {
+                // The two parent node is not the same node
+                let mut old_entry = old_parent_node.get_dir_data_mut().remove(old_name).unwrap_or_else(||unreachable!("Impossible case when exchange,\
+                                                                                                                                    as {old_name} under {old_parent} is checked to be existed."));
+                let mut new_entry = new_parent_node.get_dir_data_mut().remove(new_name).unwrap_or_else(||unreachable!("Impossible case when exchange,\
+                                                                                                                                    as {new_name} under {new_parent} is checked to be existed."));
+                old_entry.set_entry_name(new_name.into());
+                new_entry.set_entry_name(old_name.into());
+                old_parent_node.insert_entry_for_rename(new_entry);
+                new_parent_node.insert_entry_for_rename(old_entry);
+
+                txn.set(
+                    &KeyType::INum2Node(old_parent),
+                    &ValueType::Node(old_parent_node.into_serial_node()),
+                );
+                txn.set(
+                    &KeyType::INum2Node(new_parent),
+                    &ValueType::Node(new_parent_node.into_serial_node()),
+                );
+            } else {
+                let mut old_entry = old_parent_node.get_dir_data_mut().remove(old_name).unwrap_or_else(||unreachable!("Impossible case when exchange,\
+                                                                                                                                    as {old_name} under {old_parent} is checked to be existed."));
+                let mut new_entry = old_parent_node.get_dir_data_mut().remove(new_name).unwrap_or_else(||unreachable!("Impossible case when exchange,\
+                                                                                                                                    as {new_name} under {new_parent} is checked to be existed."));
+
+                old_entry.set_entry_name(new_name.into());
+                new_entry.set_entry_name(old_name.into());
+                old_parent_node.insert_entry_for_rename(new_entry);
+                old_parent_node.insert_entry_for_rename(old_entry);
+
+                txn.set(
+                    &KeyType::INum2Node(old_parent),
+                    &ValueType::Node(old_parent_node.into_serial_node()),
+                );
+            }
+
+            let success = txn.commit().await?;
+
+            if success {
+                break Ok(());
+            }
+            debug!("exchange() found transaction commit failed when exchange, retrying...");
```

Comment:
This loop potentially become an infinite loop in some cases. We should limit the retry times 🤔 ?

Reply:
The commit operation fails, if the key is updated after gotten and before committed, right? Such cases may happen in a concurrent write scene, but would it become an infinite loop? It does seem to be like "starving", but I am sorry that I can't think of such a case in normal scene.

Besides, if we should limit the retry times, what type of error should it return when it hit the limit? Is it `TransactionRetryLimitExceededErr`?

Reply:
The probability of this case occurring is very,very slow, but there always needs to be a limit for some such rare case 🥰

Reply:
Okey, thank you for your suggestion, it's a really significant lesson for me. And it's fixed now.